mm2src/mm2_core/src/sql_connection_pool.rs

```diff
@@ -0,0 +1,340 @@
+        }
+
+        // Connection pool is not already initialized, create new connection pool.
+        let conn = Self::open_connection(sqlite_file_path);
+        let connections = Arc::new(RwLock::new(HashMap::from([(db_id, conn)])));
+        let db_root = ctx.conf["dbdir"].as_str();
+        try_s!(ctx.sqlite_conn_pool.pin(Self {
+            connections,
+            rmd160_hex: ctx.rmd160_hex(),
+            shared_db_id: hex::encode(*ctx.shared_db_id()),
+            db_root: db_root.map(|d| d.to_owned())
+        }));
+
+        Ok(())
+    }
+
+    /// Test method for initializing a single-user database connection in-memory.
+    pub fn init_test(ctx: &MmCtx) -> Result<(), String> { Self::init_impl_test(ctx, None, DbIdConnKind::Single) }
+
+    /// Test method for initializing a shared database connection in-memory.
+    pub fn init_shared_test(ctx: &MmCtx) -> Result<(), String> { Self::init_impl_test(ctx, None, DbIdConnKind::Shared) }
+
+    /// Internal test implementation to initialize a database connection in-memory.
+    fn init_impl_test(ctx: &MmCtx, db_id: Option<&str>, kind: DbIdConnKind) -> Result<(), String> {
+        let db_id = Self::db_id_from_ctx(ctx, db_id, &kind);
+        if let Some(pool) = ctx.sqlite_conn_pool.as_option() {
+            let connection = Arc::new(Mutex::new(Connection::open_in_memory().unwrap()));
+            let mut pool = pool.connections.write().unwrap();
+            pool.insert(db_id, connection);
+
+            return Ok(());
+        }
+
+        let connection = Arc::new(Mutex::new(Connection::open_in_memory().unwrap()));
+        let connections = Arc::new(RwLock::new(HashMap::from([(db_id, connection)])));
+        let db_root = ctx.conf["dbdir"].as_str();
+        try_s!(ctx.sqlite_conn_pool.pin(Self {
+            connections,
+            rmd160_hex: ctx.rmd160_hex(),
+            shared_db_id: hex::encode(*ctx.shared_db_id()),
+            db_root: db_root.map(|d| d.to_owned())
+        }));
+
+        Ok(())
+    }
+
+    /// Retrieves a single-user connection from the pool.
+    pub fn sqlite_conn(&self, db_id: Option<&str>) -> Arc<Mutex<Connection>> {
+        self.sqlite_conn_impl(db_id, DbIdConnKind::Single)
+    }
+
+    /// Retrieves a shared connection from the pool.
+    pub fn sqlite_conn_shared(&self, db_id: Option<&str>) -> Arc<Mutex<Connection>> {
+        self.sqlite_conn_impl(db_id, DbIdConnKind::Shared)
+    }
+
+    /// Internal implementation to retrieve or create a connection.
+    fn sqlite_conn_impl(&self, db_id: Option<&str>, kind: DbIdConnKind) -> Arc<Mutex<Connection>> {
+        let db_id = self.db_id(db_id, &kind);
+        let connections = self.connections.read().unwrap();
+        if let Some(connection) = connections.get(&db_id) {
+            return Arc::clone(connection);
+        }
+        drop(connections);
+
+        let mut connections = self.connections.write().unwrap();
+        let sqlite_file_path = self.db_dir(&db_id).join(match kind {
```

Comment:
there are several code parts where `sqlite_file_path` is initialised like this.
Maybe it could be a function.